BaseClasses.py

```diff
@@ -681,19 +681,57 @@ def can_reach_region(self, spot: str, player: int) -> bool:
+        if key_only:
+            def event_filter(location: Location):
+                return location.item is not None and getattr(location.item, "locked_dungeon_item", False)
+        else:
+            def event_filter(location: Location):
+                return (location.advancement and location not in self.events
+                        or getattr(location.item, "locked_dungeon_item", False))
+        if locations is None:
+            # `self.multiworld.get_filled_locations(player)` is avoided because it first iterates into a list and also
+            # because `location.advancement` in `event_filter` also checks for `location.item is not None`.
+            events_per_player = {player: set(filter(event_filter, locations_dict.values()))
+                                 for player, locations_dict in self.multiworld.regions.location_cache.items()}
+        else:
+            # Can't iterate self.multiworld.player_ids because it doesn't include the extra player IDs used for item
+            # links, so iterate the keys of the location_cache instead.
+            events_per_player = {player: set() for player in self.multiworld.regions.location_cache.keys()}
+            for location in locations:
+                if event_filter(location):
+                    events_per_player[location.player].add(location)
+
+        # Remove empty sets to reduce the number of iterations required to iterate through `events_per_player.items()`.
+        for player, events in list(events_per_player.items()):
+            if not events:
+                del events_per_player[player]
+
+        # The first iteration must check the locations for all players because it is not known which players might have
+        # reachable locations.
+        players_to_check = set(self.multiworld.regions.location_cache.keys())
+        while players_to_check:
+            # Iterable of all locations belonging to the players to check.
+            loc_gen: Iterable[Location] = itertools.chain.from_iterable(
+                locations for player, locations in events_per_player.items() if player in players_to_check)
+            # Iterate reachable events into a list so that the sets in `events_per_player` can be modified while
+            # iterating through the reachable events.
+            reachable_events = [location for location in loc_gen if location.can_reach(self)]
+            # Clear the players to check to prepare for the next iteration.
+            players_to_check.clear()
             for event in reachable_events:
+                events_per_player[event.player].remove(event)
                 self.events.add(event)
-                assert isinstance(event.item, Item), "tried to collect Event with no Item"
-                self.collect(event.item, True, event)
+                item = event.item
+                assert isinstance(item, Item), "tried to collect Event with no Item"
+                changed = self.collect(item, True, event)
```

Comment:
It turns out that `self.collect(item, True, event)` always returns `True` and always adds the item into `self.prog_items` (ignoring the possibility of an overridden `World.collect()` that can return `True` without adding the item into `self.prog_items`).

However, the only non-advancement items that can be placed at this point are from a world's pre-fill, so checking each collected item for `if item.advancement:` to determine if `item.player`'s locations should be included in the next iteration might actually be slower than just assuming that every item collected in the sweep is advancement.

Reply:
that's because you're filtering to `location.advancement`. if an item is progression it *should* alter state and return True (i say should because the PR to enforce this isn't merged). `assert changed` would probably be good to have

Reply:
`self.collect(item=item, event=True, location=event)` *always* returns `True` because it has
```py
        if not changed and event:
            self.prog_items[item.player][item.name] += 1
            changed = True
        [...]
        return changed
```

Locations with non-advancement items are only sneaking into the locations for the `key_only=True` case, which appears to be a symptom of #3680 and I think makes it clear that it is a bug.

Reply:
right, but i'm of the opinion we should remove the `event` arg from `state.collect`. i don't think there's any reason we should want/allow items to be added to state if they aren't progression, and for custom state modification we already have the overridable `world.collect`